Some accuracy lints for floating point operations

thiagoarrais opened the pull request:
This will add some lints for accuracy on floating point operations suggested by @clarfon in #2040 (fixes #2040).

These are the remaining lints:

- [x] x.powi(2) => x * x
- [x] x.logN() / y.logN() => x.logbase(y)
- [x] x.logbase(E) => x.log()
- [x] x.logbase(10) => x.log10()
- [x] x.logbase(2) => x.log2().
- [x] x * PI / 180 => x.to_radians()
- [x] x * 180 / PI => x.to_degrees()
- [x] (x + 1).log() => x.log_1p()
- [x] sqrt(x * x + y * y) => x.hypot(y)

changelog: Included some accuracy lints for floating point operations

flip1995 commented:
LGTM. Please read up in #4897 on the discussion of naming these lints and which lint should apply to which float expression.

thiagoarrais commented:
I'll make a note to review which lints goes into `IMPRECISE_FLOPS` or `SUBOPTIMAL_FLOPS` before removing the WIP marker. I'm not really sure which one goes where right now to be honest.

About naming, I assume I'm ok if I stick to those two names?

flip1995 commented:
You can just put the different cases where you think they fit best (and note that in the PR body for easier review please). The names SGTM.

thiagoarrais commented:
I'm trying to make `sqrt(x * x + y * y) => x.hypot(y)` but bumping into the error "could not replace range (...) maybe parts of it were already replaced". AFAIK, that is happening because `x * x + y * y` already gets fixed to `x.mul_add(x, x * y)`. Any tips on how to avoid that? Should I change the mul_add lint to skip that case?

flip1995 commented:
I think making `mul_add` skip the case is the best solution here.

thiagoarrais commented:
Still working on this. Quick question: how do you folks usually deal with cross lint errors?

I've got a problem because the lint for `x.powi(2) + y` produces `x * x + y`, which is rejected by the mul_add lint. It wants the expression to be `x.mul_add(x, y)`.

flip1995 commented:
If a lint suggests code that again gets linted by another lint, that is **not** a problem.

Only if 2 lints trigger on the **same** code simultaneously is a problem. In those cases one lint should take preference.

thiagoarrais commented:
Should I do anything differently here, then? I'm getting the error below when I run `TESTNAME=floating_point_powi cargo uitest` with my current code:

```
error: failed to compile fixed code
status: exit code: 1
command: "/data/rust-clippy/target/debug/clippy-driver" "tests/ui/floating_point_powi.fixed" "-L" "/data/rust-clippy/target/debug/test_build_base" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/data/rust-clippy/target/debug/test_build_base/floating_point_powi.stage-id" "-L" "/data/rust-clippy/target/debug/deps" "-L" "/data/rust-clippy/target/debug/deps" "-Dwarnings" "-Zui-testing" "--extern" "regex=/data/rust-clippy/target/debug/deps/libregex-c986894c984c8eac.rlib" "--extern" "serde=/data/rust-clippy/target/debug/deps/libserde-bdb8a48e18d84a1f.rlib" "--extern" "clippy_lints=/data/rust-clippy/target/debug/deps/libclippy_lints-0033f303daa204f4.rlib" "-L" "/data/rust-clippy/target/debug/test_build_base/floating_point_powi.stage-id.aux"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"multiply and add expressions can be calculated more efficiently and accurately","code":{"code":"clippy::suboptimal_flops","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/floating_point_powi.fixed","byte_start":192,"byte_end":203,"line_start":11,"line_end":11,"column_start":13,"column_end":24,"is_primary":true,"text":[{"text":"    let _ = (x * x + y).sqrt();","highlight_start":13,"highlight_end":24}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"`-D clippy::suboptimal-flops` implied by `-D warnings`","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"consider using","code":null,"level":"help","spans":[{"file_name":"tests/ui/floating_point_powi.fixed","byte_start":192,"byte_end":203,"line_start":11,"line_end":11,"column_start":13,"column_end":24,"is_primary":true,"text":[{"text":"    let _ = (x * x + y).sqrt();","highlight_start":13,"highlight_end":24}],"label":null,"suggested_replacement":"x.mul_add(x, y)","suggestion_applicability":"MachineApplicable","expansion":null}],"children":[],"rendered":null}],"rendered":"error: multiply and add expressions can be calculated more efficiently and accurately\n  --> tests/ui/floating_point_powi.fixed:11:13\n   |\nLL |     let _ = (x * x + y).sqrt();\n   |             ^^^^^^^^^^^ help: consider using: `x.mul_add(x, y)`\n   |\n   = note: `-D clippy::suboptimal-flops` implied by `-D warnings`\n\n"}
{"message":"multiply and add expressions can be calculated more efficiently and accurately","code":{"code":"clippy::suboptimal_flops","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/floating_point_powi.fixed","byte_start":224,"byte_end":235,"line_start":12,"line_end":12,"column_start":13,"column_end":24,"is_primary":true,"text":[{"text":"    let _ = (x + y * y).sqrt();","highlight_start":13,"highlight_end":24}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"consider using","code":null,"level":"help","spans":[{"file_name":"tests/ui/floating_point_powi.fixed","byte_start":224,"byte_end":235,"line_start":12,"line_end":12,"column_start":13,"column_end":24,"is_primary":true,"text":[{"text":"    let _ = (x + y * y).sqrt();","highlight_start":13,"highlight_end":24}],"label":null,"suggested_replacement":"y.mul_add(y, x)","suggestion_applicability":"MachineApplicable","expansion":null}],"children":[],"rendered":null}],"rendered":"error: multiply and add expressions can be calculated more efficiently and accurately\n  --> tests/ui/floating_point_powi.fixed:12:13\n   |\nLL |     let _ = (x + y * y).sqrt();\n   |             ^^^^^^^^^^^ help: consider using: `y.mul_add(y, x)`\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
```

flip1995 commented:
Just allow the other lint in this test file (or just for this test case).

flip1995 commented:
FYI: since this is a big PR, which seems to be under active development, I won't review every change here (At least as long as there is WIP in the title). @thiagoarrais if you have questions or want to have a review anyway, feel free to ask anytime though.

thiagoarrais commented:
No problem! Thanks for the feedback!

thiagoarrais commented:
I'm removing the WIP marker as this is ready for review.

Some of the proposed lints were already covered in master:

* x.logbase(E) => x.log()
* x.logbase(10) => x.log10()
* x.logbase(2) => x.log2()
* (x + 1).log() => x.log_1p()

And here is a summary of which lint went into each group. It wasn't really clear to me what should go where. Please let me know if any of these are in the wrong drawer.

| Lint | Lint group |
|-------------|-----------|
|x.powi(2) => x * x | `imprecise_flops` |
|x.logN() / y.logN() => x.logbase(y) | `suboptimal_flops` |
|x * PI / 180 => x.to_radians() | `imprecise_flops` |
|x * 180 / PI => x.to_degrees() | `imprecise_flops` |
|sqrt(x * x + y * y) => x.hypot(y) | `imprecise_flops` |

clarfon commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -293,6 +293,92 @@ fn check_powf(cx: &LateContext<'_, '_>, expr: &Expr<'_>, args: &[Expr<'_>]) {
+        let (lint, help, suggestion) = match value {
+            Int(2) => (
+                IMPRECISE_FLOPS,
+                "square can be computed more accurately",
```
Is this really more accurate, or just more efficient?

thiagoarrais replied:
I don't have a definitive answer here. I have been assuming all these were precision lints just because the original issue talked about accuracy suggestions. I'm open to reclassifing this under `SUBOPTIMAL_FLOPS`. What do you think?

clarfon replied:
Yeah I'd prefer that. They should generate the same answer, just one is more specialised and thus faster

thiagoarrais replied:
I'm trying to reclassify this one, but keep hitting into `failed to compile fixed code`.

The fix is generating code that gets linted by the `mul_add` lint for the `(x + y.powi(2)).sqrt()` case and I don't see how to allow mul_add while still fixing powi because now they're both under `suboptimal-flops`.

thiagoarrais replied:
Update: Managed to get past this one by changing powi to generate code that does not trigger mul_add.

clarfon commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -479,6 +573,99 @@ fn check_custom_abs(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+                cx,
+                SUBOPTIMAL_FLOPS,
+                expr.span,
+                "division of logarithms can be calculated more efficiently and accurately",
```
Isn't the resulting code the same, just, it looks cleaner this way? Maybe word this as "more clearly" while keeping the same lint group?

thiagoarrais replied:
😮

Didn't actually realize that. Will change wording.

clarfon commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -479,6 +573,99 @@ fn check_custom_abs(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+                    cx,
+                    IMPRECISE_FLOPS,
+                    expr.span,
+                    "conversion to degrees can be done more accurately",
```
Feel like accuracy is debatable here, kind of want to just call this suboptimal instead

thiagoarrais replied:
Yup. Same as above. Will put this under suboptimal. Just let me know about the previous one.

clarfon commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -479,6 +573,99 @@ fn check_custom_abs(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+                    cx,
+                    IMPRECISE_FLOPS,
+                    expr.span,
+                    "conversion to radians can be done more accurately",
```
Ditto here

thiagoarrais replied:
:+1:

clarfon commented on tests/ui/floating_point_hypot.fixed:
```diff
@@ -0,0 +1,13 @@
+    let _ = (x + 1f32).hypot(y);
+    let _ = x.hypot(y);
+    // Cases where the lint shouldn't be applied
+    let _ = x.mul_add(x, y * y).sqrt();
```
Why don't we want to include this case? Or is this just a TODO for the future?

thiagoarrais replied:
I've started trying to include this, but ended up giving up the added complexity. Left it as a TODO instead.

clarfon replied:
Makes sense. Maybe add that in the comment here so it's clear it was just left out of the current impl

clarfon commented on clippy_lints/src/floating_point_arithmetic.rs:
```diff
@@ -479,6 +573,99 @@ fn check_custom_abs(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
+    }
+}
+
+fn check_radians(cx: &LateContext<'_, '_>, expr: &Expr<'_>) {
```
Do we have any checks for constants equivalent to π/180 and it's reciprocal? I know that the approx constant lint exists and maybe that should trigger for constants that combine the two and encourage using these methods instead.

Maybe here we should also have a check for those constants.

thiagoarrais replied:
I can add that if someone points me in the right direction.

thiagoarrais commented:
With this latest commit I think I've addressed all comments, expect for [the last one](https://github.com/rust-lang/rust-clippy/pull/5443#pullrequestreview-430235573). Please take a look and let me know if you need anything else.

clarfon commented:
To be fair, that last comment can easily be left as a TODO, I was mostly mentioning it to be thorough. Awesome work as it is. :)

flip1995 commented:
I agree with @clarfon, awesome work! Can you leave a `TODO` comment in the code for [this](https://github.com/rust-lang/rust-clippy/pull/5443#discussion_r439849012).

Otherwise, I think this should be ready to go 🚀

bors commented:
:umbrella: The latest upstream changes (presumably #5763) made this pull request unmergeable. Please resolve the merge conflicts.

thiagoarrais commented:
Added `TODO` comment and rebased. This should be good to go.

flip1995 commented:
@bors r+

bors commented:
:pushpin: Commit 3065201eb3a0c4976de7ef5b5b924afde1b18325 has been approved by `flip1995`

<!-- @bors r=flip1995 3065201eb3a0c4976de7ef5b5b924afde1b18325 -->
<!-- homu: {"type":"Approved","sha":"3065201eb3a0c4976de7ef5b5b924afde1b18325","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 3065201eb3a0c4976de7ef5b5b924afde1b18325 with merge 3749ea4c25180a0d7d53785abce9334fa03a42ea...
<!-- homu: {"type":"BuildStarted","head_sha":"3065201eb3a0c4976de7ef5b5b924afde1b18325","merge_sha":"3749ea4c25180a0d7d53785abce9334fa03a42ea"} -->

flip1995 commented:
@bors rollup=iffy

flip1995 commented:
@bors retry (yeet)

bors commented:
:hourglass: Testing commit 3065201eb3a0c4976de7ef5b5b924afde1b18325 with merge 52f7940154c70f4d4c09c58cb59d697e115d63f5...
<!-- homu: {"type":"BuildStarted","head_sha":"3065201eb3a0c4976de7ef5b5b924afde1b18325","merge_sha":"52f7940154c70f4d4c09c58cb59d697e115d63f5"} -->

flip1995 commented:
@bors retry (yeet yeet)

bors commented:
:hourglass: Testing commit 3065201eb3a0c4976de7ef5b5b924afde1b18325 with merge 4b8700879c59d1673dc36ea16aff3020af23dcc5...
<!-- homu: {"type":"BuildStarted","head_sha":"3065201eb3a0c4976de7ef5b5b924afde1b18325","merge_sha":"4b8700879c59d1673dc36ea16aff3020af23dcc5"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #5792) made this pull request unmergeable. Please resolve the merge conflicts.
